File: system-test/java/src/main/java/com/optivem/shop/testkit/common/ResultAssert.java
package com.optivem.shop.testkit.common;

import org.assertj.core.api.AbstractAssert;

public class ResultAssert<T, E> extends AbstractAssert<ResultAssert<T, E>, Result<T, E>> {
    private ResultAssert(Result<T, E> actual) {
        super(actual, ResultAssert.class);
    }

    public static <T, E> ResultAssert<T, E> assertThatResult(Result<T, E> actual) {
        return new ResultAssert<>(actual);
    }

    public ResultAssert<T, E> isSuccess() {
        isNotNull();
        if (!actual.isSuccess()) {
            failWithMessage("Expected result to be success but was failure with error: %s", actual.getError());
        }
        return this;
    }

    public ResultAssert<T, E> isFailure() {
        isNotNull();
        if (!actual.isFailure()) {
            failWithMessage("Expected result to be failure but was success");
        }
        return this;
    }
}

File: system-test/typescript/src/testkit/common/result-assert.ts
import type { Result } from './result.js';

export class ResultAssert<T, E> {
  constructor(private readonly result: Result<T, E>) {}

  isSuccess(): ResultAssert<T, E> {
    if (!this.result.success) {
      throw new Error(
        `Expected success but was failure: ${JSON.stringify(this.result.error)}`
      );
    }
    return this;
  }

  isFailure(): ResultAssert<T, E> {
    if (this.result.success) {
      throw new Error(
        `Expected failure but was success: ${JSON.stringify(this.result.value)}`
      );
    }
    return this;
  }

  getValue(): T {
    if (!this.result.success) {
      throw new Error(
        `Cannot get value from failure: ${JSON.stringify(this.result.error)}`
      );
    }
    return this.result.value;
  }

  getError(): E {
    if (this.result.success) {
      throw new Error(
        `Cannot get error from success: ${JSON.stringify(this.result.value)}`
      );
    }
    return this.result.error;
  }
}

export function assertThatResult<T, E>(result: Result<T, E>): ResultAssert<T, E> {
  return new ResultAssert(result);
}
